### PR TITLE
[Reaper: reap stale merge clones and dev homes](2) Week-old dev homes

### PR DESCRIPTION
**Pull request by EdbertChan**
## Summary

Each source checkout that runs Invoker gets its own home under `~/.invoker/dev`. Nothing reaps these homes, and one Mac held 13 of them at 119 GB.

The reaper worker now reaps a dev home when no file anywhere inside it changed for 7 days and every lock pid it records has exited.

Locks include the owner lock directories and worker lock files under `locks/`, written at `packages/execution-engine/src/worker-lock.ts:130`.

A dev home whose age, lock file, or pid cannot be read is kept, logged, and counted as unchecked in the reaper pass summary.

The production home, its database, and its backups sit outside `dev` and are never read or reaped by this rule.

## Review Claim

The reaper reaps a dev home only when it is older than 7 days and no process named in its lock files is running.

## Review Lane

behavior

## Review Unit

routing

## Safety Invariant

The reaper reaps a dev home only when nothing in it changed for 7 days and the process in its lock file is gone. It never touches the live database.

## Slice Rationale

Dev homes use a lock pid as their in-use signal, unlike merge clones in the previous slice, so this rule is reviewed on its own.

## Non-goals

- Does not reap the separate `~/.invoker/test` home.
- Does not reap dev homes on remote SSH targets.
- Does not turn the reaper on; it still starts only when `reaper.enabled` is true.

## Test Plan

<details>
<summary>Test Plan</summary>

- [x] `cd packages/execution-engine && npx vitest run src/__tests__/reaper-reclaim.test.ts src/__tests__/reaper-worker.test.ts src/__tests__/reaper-worktree-responsiveness.test.ts` (before the change: the 3 new dev-home cases failed; after: 27 passed. Review follow-up: 3 more cases for nested files, worker lock files, and malformed pids failed first; after: 30 passed)
- [x] `pnpm run check:types`
- [x] `node scripts/check-added-comments.mjs`
- [x] `node scripts/check-silent-catches.mjs`

</details>

## Revert Plan

<details>
<summary>Revert Plan</summary>

- Safe to revert? Yes
- Revert command: `git revert <sha>`
- Post-revert steps: None
- Data migration? No

</details>

Depends-On: #12304

---

> [!NOTE]
> **Medium Risk**
> Recursive deletion of dev home directories is irreversible if eligibility checks misfire; mitigations include path guards, lock/PID checks, and conservative unchecked handling.
> 
> **Overview**
> Adds **`reapStaleDevelopmentHomes`** so the reaper can delete abandoned per-checkout homes under **`~/.invoker/dev`** after **7 days** with no in-tree activity (newest mtime walk) and **no live process** from directory **`*.lock/pid`** files or JSON **`locks/*.lock`** entries.
> 
> Homes that are too new, still locked by a running PID, recently touched anywhere inside, or **unreadable** (age/locks) are **kept**; unreadable ones are listed in **`unchecked`** and surfaced in pass logs/summary. Production assets at the invoker home root (e.g. root **`invoker.db`**) are outside **`dev`** and not targeted.
> 
> The reaper worker runs this step each tick (injectable **`reapDevHomes`** seam), extends the pass **summary**, stores **`devHomeResult`** in the decision payload, and can mark the pass **failed** on dev-home cleanup errors (same pattern as merge clones). Tests cover reap/keep edge cases and worker integration.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit aaad32257bf40d6fd0f28e1e649d46b3c74a7fff. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by EdbertChan** [comment]
This pull request is part of a [Mergify stack](https://docs.mergify.com/stacks/):

| # | Pull Request | Link | |
|--:|---|---|---|
| 1 | Reaper removes stale merge clones no unfinished task uses | [#12304](https://github.com/Neko-Catpital-Labs/Invoker/pull/12304) |  |
| 2 | Reaper removes week-old dev homes with no running process | [#12305](https://github.com/Neko-Catpital-Labs/Invoker/pull/12305) | 👈 |

**Comment by EdbertChan** [comment]
Invoker repair dispatch was not acknowledged. The request was recorded as pending, then closed without consuming the code-repair retry budget; automatic retry remains enabled for current head 25d0fe2.

**Comment by EdbertChan** [comment]
### Revision history

| # | Type | Changes | Reason | Date |
|---|------|---------|--------|------|
| 1 | initial | `25d0fe2` |  | 2026-09-13 19:26 UTC |
| 2 | content | [`25d0fe2 → aaad322`](https://github.com/Neko-Catpital-Labs/Invoker/compare/25d0fe2be587c9a39e223d247b37fa6c4b58cf70...e472f9595a1ad2064c30130b154455c4f5f53496) ([raw](https://github.com/Neko-Catpital-Labs/Invoker/compare/25d0fe2be587c9a39e223d247b37fa6c4b58cf70...aaad32257bf40d6fd0f28e1e649d46b3c74a7fff)) |  | 2026-09-13 19:26 UTC |

File: packages/execution-engine/src/workers/reaper-reclaim.ts [hunk 1]
```diff
@@ -613,6 +616,134 @@ export async function reapStaleMergeClones(opts: {
...
+    const pidPath = join(devHome, name, 'pid');
+    if (!existsSync(pidPath)) continue;
+    let raw: string;
+    try {
+      raw = readFileSync(pidPath, 'utf8');
+    } catch (err) {
+      return { state: 'unreadable', detail: `${pidPath}: ${errorDetail(err)}` };
+    }
+    const pid = Number.parseInt(raw.trim(), 10);
+    if (!Number.isNaN(pid) && isProcessAlive(pid)) return { state: 'alive', pid, lock: name };
```
[the remark this answers is not in the document]
Comment (EdbertChan) [reply packages/execution-engine/src/workers/reaper-reclaim.ts]: Fixed in aaad322.

The worker lock format is confirmed at `packages/execution-engine/src/worker-lock.ts:130-137`: `<home>/locks/worker-<kind>.lock` holds JSON with a `pid`. The old check only looked at `*.lock/pid` directories, so a live worker in an old dev home would have been missed.

`findLockHolder` now also reads every `locks/*.lock` file (`packages/execution-engine/src/workers/reaper-reclaim.ts:707-729` at aaad322). A running pid keeps the home. A file that cannot be parsed, or a pid that is not a positive safe integer, marks the home unchecked instead of deletable.

New case `keeps an old dev home whose worker lock file names a running process` (`packages/execution-engine/src/__tests__/reaper-reclaim.test.ts:425`) failed before the fix and passes after:

```
× reapStaleDevelopmentHomes > keeps an old dev home whose worker lock file names a running process
      Tests  3 failed | 3 passed | 19 skipped (25)
```

```
 Test Files  3 passed (3)
      Tests  30 passed (30)
```

File: packages/execution-engine/src/workers/reaper-reclaim.ts [hunk 2]
```diff
@@ -613,6 +616,134 @@ export async function reapStaleMergeClones(opts: {
...
+  userHome?: string;
+  nowMs?: number;
+  isProcessAlive?: (pid: number) => boolean;
+}): Promise<StaleDevelopmentHomeReapResult> {
+  const userHome = opts.userHome ?? homedir();
+  const home = expandTildeHome(opts.invokerHome, userHome);
+  if (!isSafeInvokerHome(home, userHome)) return { ok: false, removed: [], unchecked: [], reason: 'path-guard' };
+
+  const devRoot = join(home, 'dev');
+  if (!isDirectory(devRoot)) return { ok: true, removed: [], unchecked: [] };
```
[the remark this answers is not in the document]
Comment (EdbertChan) [reply packages/execution-engine/src/workers/reaper-reclaim.ts]: Fixed in aaad322.

A new `pathKind` helper (`packages/execution-engine/src/workers/reaper-reclaim.ts:637`) separates a missing path from a path that could not be checked.

- Dev root (`reaper-reclaim.ts:746`): a failed check now returns `ok: false` with a `cleanup-error` reason and an error log. A missing or non-directory root still returns an empty success.
- Each dev home (`reaper-reclaim.ts:769-771`): a failed check now adds the home to `unchecked` with a warning, instead of skipping it silently.

This fix has no dedicated test. A permission-denied `stat` cannot be reproduced reliably when CI runs as root, so the change rests on code review plus the full reaper suite still passing:

```
 Test Files  3 passed (3)
      Tests  30 passed (30)
types_exit=0
```
